modules/transport-netty4/src/main/java/org/elasticsearch/transport/netty4/Netty4WriteThrottlingHandler.java

```diff
@@ -43,7 +46,54 @@ public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
         assert msg instanceof ByteBuf;
         assert Transports.assertDefaultThreadContext(threadContext);
         assert Transports.assertTransportThread();
-        final boolean queued = queuedWrites.offer(new WriteOperation((ByteBuf) msg, promise));
+        final ByteBuf buf = (ByteBuf) msg;
+        if (ctx.channel().isWritable() && currentWrite == null && queuedWrites.isEmpty()) {
+            // nothing is queued for writing and the channel is writable, just pass the write down the pipeline directly
+            if (buf.readableBytes() > MAX_BYTES_PER_WRITE) {
+                writeInSlices(ctx, promise, buf);
+            } else {
+                ctx.write(msg, promise);
+            }
+        } else {
+            queueWrite(buf, promise);
+        }
+    }
+
+    /**
+     * Writes slices of up to the max write size until the channel stops being writable or the message has been written in full.
+     */
+    private void writeInSlices(ChannelHandlerContext ctx, ChannelPromise promise, ByteBuf buf) {
+        while (true) {
```

Comment:
I guess we don't really need a loop here since this will always just run a single time because the outbound buffer on the Netty end is less than `MAX_BYTES_PER_WRITE` in all cases today. We do have the same kind of loop in `doFlush` though, so I made it a loop here as well for consistency. Also, this makes the logic stable in case the outbound buffer size ever changes.